quic/s2n-quic-core/src/crypto/tls/offload.rs

```diff
@@ -0,0 +1,567 @@
+        }
+    }
+}
+
+struct ThreadWaker(thread::Thread);
+
+impl Wake for ThreadWaker {
+    fn wake(self: Arc<Self>) {
+        self.0.unpark();
+    }
+}
+
+impl<E: tls::Endpoint> tls::Endpoint for OffloadEndpoint<E> {
+    type Session = OffloadSession<<E as tls::Endpoint>::Session>;
+
+    fn new_server_session<Params: s2n_codec::EncoderValue>(
+        &mut self,
+        transport_parameters: &Params,
+    ) -> Self::Session {
+        OffloadSession::new(
+            self.inner.new_server_session(transport_parameters),
+            &mut self.new_session,
+            self.remote_thread_waker.clone(),
+        )
+    }
+
+    fn new_client_session<Params: s2n_codec::EncoderValue>(
+        &mut self,
+        transport_parameters: &Params,
+        server_name: application::ServerName,
+    ) -> Self::Session {
+        OffloadSession::new(
+            self.inner
+                .new_client_session(transport_parameters, server_name),
+            &mut self.new_session,
+            self.remote_thread_waker.clone(),
+        )
+    }
+
+    fn max_tag_length(&self) -> usize {
+        self.inner.max_tag_length()
+    }
+}
+
+#[derive(Debug)]
+pub struct OffloadSession<S: tls::Session> {
+    // Inner is none while remote thread has the session
+    inner: Option<S>,
+    is_poll_done: Option<Result<(), crate::transport::Error>>,
+    pending_requests: UnboundedReceiver<Request<S>>,
+    waker: Waker,
+}
+
+impl<S: tls::Session> OffloadSession<S> {
+    fn new(
+        inner: S,
+        new_session: &mut UnboundedSender<(S, UnboundedSender<Request<S>>)>,
+        remote_thread: Waker,
+    ) -> Self {
+        // Channel to pass requests from remote TLS thread to main thread
+        let (tx, rx) = futures_channel::mpsc::unbounded::<Request<S>>();
+
+        // Send the session to the TLS thread. It will pass it back when the handshake has finished.
+        let _ = new_session.unbounded_send((inner, tx));
+
+        Self {
+            pending_requests: rx,
+            waker: remote_thread,
+            is_poll_done: None,
+            inner: None,
+        }
+    }
+}
+
+impl<S: tls::Session> tls::Session for OffloadSession<S> {
+    #[inline]
+    fn poll<W>(&mut self, context: &mut W) -> Poll<Result<(), transport::Error>>
+    where
+        W: tls::Context<Self>,
+    {
+        if let Some(finished) = self.is_poll_done {
+            return Poll::Ready(finished);
+        }
+        // This will wake up the TLS remote thread
+        self.waker.wake_by_ref();
+
+        loop {
+            let mut cx = Context::from_waker(context.waker());
+
+            let req = match Pin::new(&mut self.pending_requests).poll_next(&mut cx) {
+                Poll::Ready(Some(request)) => request,
+                Poll::Ready(None) => {
+                    return Poll::Ready(Err(crate::transport::Error::INTERNAL_ERROR
+                        .with_reason("offloaded crypto session finished without sending Done")))
```

Comment:
this would only happen if there was some bug in the impl? Or is there an expected case this would happen?